google/resource_cloudiot_registry.go

```diff
@@ -0,0 +1,411 @@
+		}
+		for _, updateMaskItem := range updateMask {
+			d.SetPartial(updateMaskItem)
+		}
+	}
+
+	d.Partial(false)
+
+	return resourceCloudiotRegistryRead(d, meta)
+}
+
+func resourceCloudiotRegistryRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	name := d.Id()
+	res, err := config.clientCloudiot.Projects.Locations.Registries.Get(name).Do()
+
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Registry %q", name))
+	}
+
+	d.Set("name", res.Id)
+
+	if res.EventNotificationConfigs != nil {
+		eventConfigs := make([]map[string]string, len(res.EventNotificationConfigs))
+		for i, config := range res.EventNotificationConfigs {
+			eventConfigs[i] = map[string]string{"pubsub_topic_name": config.PubsubTopicName}
+		}
+		d.Set("event_notification_configs", eventConfigs)
+	}
+	// to keep the data model lean only changes are pushed if not the default is
+	// returned or a config exists for state notification, mqtt and http config.
+	if res.StateNotificationConfig != nil {
+		pubsubTopicName := res.StateNotificationConfig.PubsubTopicName
+		_, hasStateConfig := d.GetOk("state_notification_config")
+		if pubsubTopicName != "" || hasStateConfig {
+			d.Set("state_notification_config",
+				map[string]string{"pubsub_topic_name": pubsubTopicName})
+		}
+	}
+	if res.MqttConfig != nil {
+		mqttState := res.MqttConfig.MqttEnabledState
+		_, hasMqttConfig := d.GetOk("mqtt_config")
+		if mqttState != mqttEnabled || hasMqttConfig {
+			d.Set("mqtt_config",
+				map[string]string{"mqtt_enabled_state": mqttState})
+		}
+	}
+	if res.HttpConfig != nil {
+		httpState := res.HttpConfig.HttpEnabledState
+		_, hasHttpConfig := d.GetOk("http_config")
+		if httpState != httpEnabled || hasHttpConfig {
+			d.Set("http_config",
+				map[string]string{"http_enabled_state": httpState})
+		}
+	}
+
+	if res.Credentials != nil {
+		credentials := make([]map[string]interface{}, len(res.Credentials))
+		for i, item := range res.Credentials {
+			pubcert := make(map[string]interface{})
+			pubcert["format"] = item.PublicKeyCertificate.Format
+			pubcert["certificate"] = item.PublicKeyCertificate.Certificate
+
+			credentials[i] = pubcert
+		}
+		d.Set("credentials", credentials)
+	}
+
+	return nil
+}
+
+func resourceCloudiotRegistryDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	name := d.Id()
+	call := config.clientCloudiot.Projects.Locations.Registries.Delete(name)
+	_, err := call.Do()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceCloudiotRegistryStateImporter(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return nil, err
+	}
+
+	region, err := getRegion(d, config)
+	if err != nil {
+		return nil, err
+	}
+
+	id := fmt.Sprintf("projects/%s/locations/%s/registries/%s", project, region, d.Id())
```

Comment:
Problem - in this mode, the user must configure the provider such that the default region and project are the right ones - d.Id() is the only thing you'll have going into the import.

Reply:
Not thought about that. I now ask for full path `projects/{project}/locations/{region}/registries/{name}` as id. (see also [registry get API](https://cloud.google.com/iot/docs/reference/rest/v1/projects.locations.registries/get)).